### ISSUE TITLE
Codeclause_Facial Emotions Detection

### ISSUE THREAD

**Issue by Shivharekanhaiya** [Shivharekanhaiya/kanhaiya-shivhare#3]
To build a project for facial emotions detection with a graphical user interface (GUI) using Python programming, machine learning, and artificial intelligence, we can utilize the following steps:

Dataset Collection: Obtain a dataset of facial images labeled with different emotions. There are various publicly available datasets, such as the "Facial Expression Recognition Challenge" (FERC) dataset or the "Extended Cohn-Kanade (CK+)" dataset.

Data Preprocessing: Clean and preprocess the dataset by resizing the images, converting them to grayscale, and normalizing the pixel values.

Feature Extraction: Extract relevant features from the preprocessed images. One popular approach is to use the "Facial Landmarks" technique, where you identify key points on the face, such as the eyes, nose, and mouth, using techniques like the "dlib" library.

Model Training: Train a machine learning or deep learning model on the preprocessed dataset. Popular choices include Convolutional Neural Networks (CNNs) or Recurrent Neural Networks (RNNs). You can use libraries like TensorFlow, Keras, or PyTorch for this purpose.

Model Evaluation: Evaluate the trained model's performance using appropriate metrics, such as accuracy, precision, recall, or F1-score. This will help you assess how well your model is able to detect facial emotions.

GUI Development: Create a GUI using a library like Tkinter, PyQt, or PySide. The GUI should provide functionality for users to upload an image or access the device's camera to capture a real-time image.

Facial Emotion Detection: Implement the facial emotion detection algorithm in the GUI. Pass the selected or captured image through the trained model to predict the emotion. Display the predicted emotion on the GUI along with the image.

Here's a high-level example demonstrating the steps above:

import tkinter as tk
from PIL import ImageTk, Image
import cv2
import numpy as np
import dlib

# Load the pre-trained facial emotion detection model
model = ...  # Load your trained model here

# Load the facial landmarks detector from dlib
detector = dlib.get_frontal_face_detector()
predictor = dlib.shape_predictor("shape_predictor_68_face_landmarks.dat")

# Create a function to detect facial emotions
def detect_emotions(image_path):
    # Load the image
    image = cv2.imread(image_path)
    
    # Convert the image to grayscale
    gray_image = cv2.cvtColor(image, cv2.COLOR_BGR2GRAY)
    
    # Detect faces in the grayscale image
    faces = detector(gray_image)
    
    # Iterate over each detected face
    for face in faces:
        # Detect facial landmarks
        landmarks = predictor(gray_image, face)
        
        # Extract features from the facial landmarks (e.g., distances between key points)
        features = extract_features(landmarks)
        
        # Preprocess the features (e.g., scaling or normalization)
        preprocessed_features = preprocess_features(features)
        
        # Make predictions using the preprocessed features
        predicted_emotion = model.predict(preprocessed_features)
        
        # Draw a rectangle around the face
        x, y, w, h = face.left(), face.top(), face.width(), face.height()
        cv2.rectangle(image, (x, y), (x+w, y+h), (0, 255, 0), 2)
        
        # Display the predicted emotion on the image
        cv2.putText(image, predicted_emotion, (x, y-10), cv2.FONT_HERSHEY_SIMPLEX, 0.9, (0, 255, 0), 2)
    
    # Convert the OpenCV image to PIL format
    image_pil = Image.fromarray(cv2.cvtColor(image, cv2.COLOR_BGR2RGB))
    
    # Create a Tkinter-compatible image
    photo = ImageTk.PhotoImage(image_pil)
    
    # Display the image in the GUI
    canvas.create_image(0, 0, image=photo, anchor=tk.NW)
    canvas.image = photo

# Create a function to handle the button click event
def browse_image():
    # Open a file dialog to select an image file
    image_path = tk.filedialog.askopenfilename()
    
    # Call the detect_emotions function with the selected image path
    detect_emotions(image_path)

# Create the main window
window = tk.Tk()
window.title("Facial Emotions Detection")

# Create a canvas to display the image
canvas = tk.Canvas(window, width=800, height=600)
canvas.pack()

# Create a button to browse for an image
browse_button = tk.Button(window, text="Browse", command=browse_image)
browse_button.pack()

# Start the Tkinter event loop
window.mainloop()
